### PR TITLE
improve disk queue performance by coalescing writes

### PR DESCRIPTION
**Pull request by leehinman**
Improves speed and lowers allocations

## What does this PR do?

Improves speed and lowers allocations in disk queue by coalescing writes.

## Why is it important?

lowers performance cost of using disk queue.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
~~- [ ] I have made corresponding changes to the documentation~~
~~- [ ] I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## How to test this PR locally

Tested with `libbeat/publisher/queue/diskqueue/benchmark_test.go`

before change
`go test -bench=. -benchtime 1x -count 10 -timeout 120m -benchmem > baseline.txt`

after change
`go test -bench=. -benchtime 1x -count 10 -timeout 120m -benchmem > gather_w.txt`

```
benchstat baseline.txt gather_w.txt
name       old time/op    new time/op    delta
1M_10-16      38.7s ± 2%     21.8s ± 0%  -43.58%  (p=0.000 n=10+7)
1M_100-16     38.7s ± 1%     21.8s ± 1%  -43.74%  (p=0.000 n=10+10)
1M_1k-16      38.6s ± 1%     21.9s ± 1%  -43.36%  (p=0.000 n=10+10)
1M_10k-16     38.7s ± 2%     21.7s ± 0%  -43.78%  (p=0.000 n=10+10)

name       old alloc/op   new alloc/op   delta
1M_10-16     3.47GB ± 0%    3.43GB ± 0%   -1.16%  (p=0.000 n=8+10)
1M_100-16    3.47GB ± 0%    3.43GB ± 0%   -1.17%  (p=0.000 n=10+8)
1M_1k-16     3.47GB ± 0%    3.43GB ± 0%   -1.18%  (p=0.000 n=10+8)
1M_10k-16    3.47GB ± 0%    3.43GB ± 0%   -1.18%  (p=0.000 n=10+10)

name       old allocs/op  new allocs/op  delta
1M_10-16      43.4M ± 0%     40.6M ± 0%   -6.32%  (p=0.000 n=8+10)
1M_100-16     43.4M ± 0%     40.6M ± 0%   -6.42%  (p=0.000 n=10+9)
1M_1k-16      43.4M ± 0%     40.6M ± 0%   -6.49%  (p=0.000 n=10+10)
1M_10k-16     43.4M ± 0%     40.6M ± 0%   -6.49%  (p=0.000 n=10+10)
```

## Related issues

- Relates elastic/elastic-agent-shipper#33

### CODE REVIEW

File: libbeat/publisher/queue/diskqueue/writer_loop.go [hunk 1]
```diff
@@ -91,9 +92,13 @@ type writerLoop struct {
 	outputFile *os.File
 
 	currentRetryInterval time.Duration
+
+	// buffer Used to gather write information so there is only one write syscall
+	buffer *bytes.Buffer
 }
 
 func newWriterLoop(logger *logp.Logger, settings Settings) *writerLoop {
+	buffer := &bytes.Buffer{}
```
Comment (cmacknz) [inline libbeat/publisher/queue/diskqueue/writer_loop.go:101]: As another minor optimization, are you able to know how big this needs to be before using it? You could make sure it only allocates once with https://pkg.go.dev/bytes#Buffer.Grow

Alternatively we could reuse it between calls with https://pkg.go.dev/bytes#Buffer.Reset but we'll hang on to the largest size buffer ever used indefinitely.

**Comment by cmacknz** [comment]
LGTM, I'll let @faec do the approving here. The test failure is fixed by https://github.com/elastic/beats/pull/31943

File: libbeat/publisher/queue/diskqueue/writer_loop.go [hunk 2]
```diff
@@ -188,22 +194,27 @@ outerLoop:
 		// The Write calls below all pass through retryWriter, so they can
 		// only return an error if the write should be aborted. Thus, all we
 		// need to do when we see an error is break out of the request loop.
-		err := binary.Write(retryWriter, binary.LittleEndian, frameSize)
+		err := binary.Write(wl.buffer, binary.LittleEndian, frameSize)
```
[the remark this answers is not in the document]
Comment (leehinman) [reply libbeat/publisher/queue/diskqueue/writer_loop.go]: updated.  thanks for catching that.

File: libbeat/publisher/queue/diskqueue/writer_loop.go (hunk 1 above)
Comment (leehinman) [reply libbeat/publisher/queue/diskqueue/writer_loop.go:101]: The writer loop is only initialized once, so this bytes.Buffer is used for the lifetime of the queue.  Unfortunately when the writer_loop is started no writes have come in yet so we don't know how big to make it.

There is a call to Buffer.Reset(), so we are keeping the largest buffer used for the lifetime of the queue.  It shouldn't be huge, the size of the largest frame, which will only have one event.

Comment (cmacknz) [reply libbeat/publisher/queue/diskqueue/writer_loop.go:101]: Thanks! I was mislead by this being in a function where I could see the lifecycle in the diff.

**Review by cmacknz — APPROVED** [review]

**Pull request merged**